File: Chapter03/Swift/WaypointList.swift
//
//  WaypointList.swift
//  EverydayDataStructures-Swift
//
//  Created by William Smith on 8/26/16.
//  Copyright © 2016 Websmiths, LLC. All rights reserved.
//

import Foundation

open class WaypointList
{
    var _route: Array = [Waypoint]()
    var _current: Waypoint?

    init() { }

    open func addWaypoints(_ waypoints: Array<Waypoint>)
    {
        _route.append(contentsOf: waypoints)
    }

    open func removeWaypoint(_ waypoint: Waypoint) -> Bool
    {
        if let index = _route.index(of: waypoint)
        {
            _route.remove(at: index)
            return true
        }
        return false
    }

    open func insertWaypoints(_ waypoints: Array<Waypoint>, before: Waypoint)
    {
        if let index = _route.index(of: before)
        {
            _route.insert(contentsOf: waypoints, at:index)
        }
        else
        {
            addWaypoints(waypoints)
        }
    }

    open func startRoute() -> Bool
    {
        if _route.count > 1
        {
            _current = startingLine()
            return moveToNextWaypoint()
        }
        return false
    }

    open func moveToNextWaypoint() -> Bool
    {
        if (_current != nil)
        {
            _current!.DeactivateWaypoint()
            if _current != self.finishLine()
            {
                let index = _route.index(of: _current!)
                _current = _route[index!+1]
                return true
            }
            return false;
        }
        return false
    }

    open func moveToPreviousWaypoint() -> Bool
    {
        if (_current != nil &&
            _current != self.startingLine())
        {
            let index = _route.index(of: _current!)
            _current = _route[index!-1]
            _current!.ReactivateWaypoint()
            return true
        }
        return false
    }

    open func startingLine() -> Waypoint
    {
        return _route.first!
    }

    open func finishLine() -> Waypoint
    {
        return _route.last!
    }

    open func currentWaypoint() -> Waypoint
    {
        return _current!
    }
}

File: Chapter03/Java/WaypointList.java
package main.java;

import java.util.List;
import java.util.LinkedList;

/**
 * Created by William Smith on 8/25/16.
 */
public class WaypointList
{
    LinkedList<Waypoint> route;
    Waypoint current;

    public WaypointList()
    {
        this.route = new LinkedList<Waypoint>();
    }

    public void AddWaypoints(List<Waypoint> waypoints)
    {
        this.route.addAll(waypoints);
    }

    public boolean RemoveWaypoint(Waypoint waypoint)
    {
        return this.route.remove(waypoint);
    }

    public void InsertWaypointsBefore(List<Waypoint> waypoints, Waypoint before)
    {
        int index = this.route.indexOf(before);
        if (index >= 0)
        {
            this.route.addAll(index, waypoints);
        }
        else
        {
            this.AddWaypoints(waypoints);
        }
    }

    public boolean StartRoute()
    {
        if (this.route.size() > 1)
        {
            this.current = this.StartingLine();
            return this.MoveToNextWaypoint();
        }
        return false;
    }

    public boolean MoveToNextWaypoint()
    {
        if (this.current != null)
        {
            this.current.DeactivateWaypoint();
            if (this.current != this.FinishLine())
            {
                int index = this.route.indexOf(this.current);
                this.current = this.route.listIterator(index).next();
                return true;
            }
            return false;
        }
        return false;
    }

    public boolean MoveToPreviousWaypoint()
    {
        if (this.current != null &&
            this.current != this.StartingLine())
        {
            int index = this.route.indexOf(this.current);
            this.current = this.route.listIterator(index).previous();
            this.current.ReactivateWaypoint();
            return true;
        }
        return false;
    }

    public Waypoint StartingLine()
    {
        return this.route.getFirst();
    }

    public Waypoint FinishLine()
    {
        return this.route.getLast();
    }

    public Waypoint CurrentWaypoint()
    {
        return this.current;
    }
}
